tools/crm_mon.c

```diff
@@ -3005,57 +3005,40 @@ print_failed_actions(FILE *stream, pe_working_set_t *data_set)
 static stonith_history_t *
 reduce_stonith_history(stonith_history_t *history)
 {
-    stonith_history_t *new = NULL, *hp, *np, *tmp;
+    stonith_history_t *new = history, *hp, *np;
 
-    for (hp = history; hp; ) {
-        for (np = new; np; np = np->next) {
-            if ((hp->state == st_done) || (hp->state == st_failed)) {
-                /* action not in progress */
-                if (safe_str_eq(hp->target, np->target) &&
-                    safe_str_eq(hp->action, np->action) &&
-                    (hp->state == np->state)) {
-                    if ((hp->state == st_done) ||
-                        safe_str_eq(hp->delegate, np->delegate)) {
-                        /* replace or purge */
-                        if (hp->completed < np->completed) {
+    if (new) {
+        hp = new->next;
+        new->next = NULL;
+
+        while (hp) {
+            stonith_history_t *hp_next = hp->next;
+
+            hp->next = NULL;
+
+            for (np = new; ; np = np->next) {
+                if ((hp->state == st_done) || (hp->state == st_failed)) {
+                    /* action not in progress */
+                    if (safe_str_eq(hp->target, np->target) &&
+                        safe_str_eq(hp->action, np->action) &&
+                        (hp->state == np->state) &&
+                        ((hp->state == st_done) ||
+                         safe_str_eq(hp->delegate, np->delegate))) {
                             /* purge older hp */
-                            tmp = hp->next;
-                            hp->next = NULL;
                             stonith_history_free(hp);
-                            hp = tmp;
                             break;
-                        }
-                        /* damn single linked list */
-                        free(hp->target);
-                        free(hp->action);
-                        free(np->origin);
-                        np->origin = hp->origin;
-                        free(np->delegate);
-                        np->delegate = hp->delegate;
-                        free(np->client);
-                        np->client = hp->client;
-                        np->completed = hp->completed;
-                        tmp = hp;
-                        hp = hp->next;
-                        free(tmp);
-                        break;
                     }
                 }
-                if (np->next) {
-                    continue;
+
+                if (!np->next) {
+                    np->next = hp;
```

Comment:
I think this will be broken if hp just matched the last entry in new and was freed. It should be sufficient to set hp = NULL after freeing it.

Reply:
Even if hp matches the last entry of new and deletes hp, I think this process will not be executed because it will be "breaked".
Is this wrong?

However, I think that setting NULL after deleting hp is safe code.

Reply:
Ah, I overlooked the break. It's fine as it is, I'll merge it.

Reply:
Hi Ken,

All right!
Thanks!